### PR TITLE
fix: initialize transactionMu correctly and ensure the transitivity of transactionMu

### PR DESCRIPTION
**Pull request by kingzytgit**
In the Transaction, there was an issue with the original code that checked if transactionMu was nil. Once entered, it couldn't exit. The problem was that Store(false) should have been Store(true). However, this approach didn't feel very reliable to me, so I removed this initialization. Instead, I added the initialization of transactionMu in each of the function to create new Adapter. Additionally, at the end of the Transaction, I continued to pass transactionMu down

### CODE REVIEW

**Comment by hsluoyz** [comment]
@MuZhou233 plz review

File: adapter.go [hunk 1]
```diff
@@ -690,31 +692,22 @@ func (a *Adapter) AddPolicies(sec string, ptype string, rules [][]string) error
 
 // Transaction perform a set of operations within a transaction
 func (a *Adapter) Transaction(e casbin.IEnforcer, fc func(casbin.IEnforcer) error, opts ...*sql.TxOptions) error {
-	// ensure the transactionMu is initialized
-	if a.transactionMu == nil {
```
Comment (MuZhou233) [inline adapter.go:694]: It would be better to keep the nil check and return error while transactionMu is nil.

File: adapter.go [hunk 2]
```diff
@@ -690,31 +692,22 @@ func (a *Adapter) AddPolicies(sec string, ptype string, rules [][]string) error
...
-		for a.muInitialize.CompareAndSwap(false, true) {
```
Comment (MuZhou233) [inline adapter.go:695]: We can delete `muInitialize` from `Adapter` now.

**Comment by hsluoyz** [comment]
@kingzyt

**Comment by kingzytgit** [comment]
I'm glad to hear there's progress on the issue! My PR is just a suggestion, to be honest, I haven't deeply considered this bug. I hope this PR is helpful for your project.

From a concurrency-safe perspective, transactionMu should not be nil from the beginning of NewAdapter, and it should not be modified. Keeping the nil check for Transaction seems unnecessary to me from a logical standpoint. In high-concurrency scenarios, if transactionMu is initially nil, and two Transaction instances are executed almost simultaneously, and both pass the transactionMu==nil check before either of them assigns a value to transactionMu, the second assignment will overwrite the first one. This can lead to a concurrency safety issue where not the same mutex protects both operations, but rather each mutex assumes it is protecting the operation.

**Comment by MuZhou233** [comment]
> if transactionMu is initially nil, and two Transaction instances are executed almost simultaneously, and both pass the transactionMu==nil check before either of them assigns a value to transactionMu, the second assignment will overwrite the first one. This can lead to a concurrency safety issue where not the same mutex protects both operations, but rather each mutex assumes it is protecting the operation.

That's exactly the original purpose of using `muInitialize`.

> Keeping the nil check for Transaction seems unnecessary to me from a logical standpoint. 

Sure it is not such neccessay. I'm thinking the `transactionMu` is newly added and there may be code somewhere didn't initialize it. A nil check could avoid potential panic. (e.g. If someone create `Adapter` without provided function, mutex would be nil after upgrade this lib)

---

The old mutex initialize logic do have concurrency safety issues. I think your solution is better.

**Comment by yuikns** [comment]
I am meeting the same issue. I noticed the previous PR has been beautifully crafted and I'm really impressed with the work done.
May I know will it merged soon? @MuZhou233

**Comment by MuZhou233** [comment]
> I am meeting the same issue. I noticed the previous PR has been beautifully crafted and I'm really impressed with the work done. May I know will it merged soon? @MuZhou233

Yes, we were waiting for @kingzytgit to make some code changes as I suggested. Since a month has passed, I will draft another PR to solve this issue.

**Comment by hsluoyz** [comment]
Replaced by: https://github.com/casbin/gorm-adapter/pull/237

**Pull request closed without merging**